quipucords/api/credential/serializer.py

```diff
@@ -23,86 +23,134 @@ def expand_filepath(filepath):
+    class NewCredentialSerializer(BaseCredentialSerializer):
+        class Meta:
+            model = Credential
+            fields = ["id", "name", "cred_type", "new_field"]
+            extra_kwargs = {
+                "new_field": {"required": True, **ENCRYPTED_FIELD_KWARGS},
+            }
+            qpc_data_sources = [DataSources.NEW_SOURCE]
+    """
 
     name = CharField(required=True, max_length=64)
-    cred_type = ValidStringChoiceField(required=False, choices=DataSources.choices)
-    username = CharField(required=False, max_length=64)
-    password = CharField(
-        required=False,
-        max_length=1024,
-        allow_null=True,
-        style={"input_type": "password"},
-    )
-    auth_token = CharField(
-        required=False,
-        max_length=1024,
-        allow_null=True,
-        style={"input_type": "password"},
-    )
-    ssh_keyfile = CharField(required=False, max_length=1024, allow_null=True)
-    ssh_passphrase = CharField(
-        required=False,
-        max_length=1024,
-        allow_null=True,
-        style={"input_type": "password"},
-    )
-    become_method = ValidStringChoiceField(
-        required=False, choices=Credential.BECOME_METHOD_CHOICES
-    )
-    become_user = CharField(required=False, max_length=64)
-    become_password = CharField(
-        required=False,
-        max_length=1024,
-        allow_null=True,
-        style={"input_type": "password"},
-    )
 
     class Meta:
         """Metadata for the serializer."""
 
         model = Credential
         fields = "__all__"
+        extra_kwargs = {
+            "password": ENCRYPTED_FIELD_KWARGS,
+            "auth_token": ENCRYPTED_FIELD_KWARGS,
+            "ssh_passphrase": ENCRYPTED_FIELD_KWARGS,
+            "become_password": ENCRYPTED_FIELD_KWARGS,
+        }
+
+    def __init__(self, instance=None, data=empty, **kwargs):
+        """Customize class initialization."""
+        if instance and isinstance(data, dict):
+            # assume cred_type == instance.cred_type if not provided
+            # this is only required to avoid breaking current functionality
+            # that was treating 'cred_type' as optional for updates and required for
+            # credential creation
+            data.setdefault("cred_type", instance.cred_type)
+        super().__init__(instance=instance, data=data, **kwargs)
+
+    def __init_subclass__(cls, **kwargs):
+        """Overload subclass creation method."""
+        super().__init_subclass__(**kwargs)
+        if not getattr(cls.Meta, "qpc_data_sources", None):
+            raise NotImplementedError(
+                "subclasses of CredentialSerializer MUST have a 'Meta' class "
+                "with 'qpc_data_sources' set as a list of DataSources."
+            )
+
+        # pylint doesn't know cls in __init_subclass__ refer to it's subclasses...
+        # pylint: disable=no-member
+        for data_source in cls.Meta.qpc_data_sources:
+            # pylint: enable=no-member
```

Comment:
Minor style question: why is this defined as a block disable instead of an inline disable?

```suggestion
        for data_source in cls.Meta.qpc_data_sources:  # pylint: disable=no-member
```